pkg/controller/deployment/deployment_controller.go

```diff
@@ -438,41 +439,143 @@ func (dc *DeploymentController) syncDeployment(key string) error {
+	if activeOrLatest := findActiveOrLatest(newRS, oldRSs); activeOrLatest != nil {
+		if activeOrLatest.Spec.Replicas == deployment.Spec.Replicas {
+			return nil
+		}
+		_, _, err := dc.scaleReplicaSetAndRecordEvent(activeOrLatest, deployment.Spec.Replicas, deployment)
+		return err
+	}
+
+	// If the new replica set is saturated, old replica sets should be fully scaled down.
+	// This case handles replica set adoption during a saturated new replica set.
+	if deploymentutil.IsSaturated(deployment, newRS) {
+		for _, old := range controller.FilterActiveReplicaSets(oldRSs) {
+			if _, _, err := dc.scaleReplicaSetAndRecordEvent(old, 0, deployment); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+
+	// There are old replica sets with pods and the new replica set is not saturated.
+	// We need to proportionally scale all replica sets (new and old) in case of a
+	// rolling deployment.
+	if deploymentutil.IsRollingUpdate(deployment) {
+		allRSs := controller.FilterActiveReplicaSets(append(oldRSs, newRS))
+		allRSsReplicas := deploymentutil.GetReplicaCountForReplicaSets(allRSs)
+
+		// Number of additional replicas that can be either added or removed from the total
+		// replicas count. These replicas should be distributed proportionally to the active
+		// replica sets. Any leftovers should be distributed as well.
+		deploymentReplicasToAdd := deployment.Spec.Replicas - allRSsReplicas
+		if deploymentReplicasToAdd == 0 {
+			return nil
+		}
+
+		// The additional replicas should be distributed proportionally amongst the active
+		// replica sets from the larger to the smaller in size replica set. Scaling direction
+		// drives what happens in case we are trying to scale replica sets of the same size.
+		// In such a case when scaling up, we should scale up newer replica sets first, and
+		// when scaling down, we should scale older replica sets first.
+		scaleUp := deploymentReplicasToAdd > 0
+		scalingOperation := "up"
+		if scaleUp {
+			sort.Sort(controller.ReplicaSetsBySizeNewer(allRSs))
+		} else {
+			sort.Sort(controller.ReplicaSetsBySizeOlder(allRSs))
+			scalingOperation = "down"
+		}
+
+		// Iterate over all active replica sets and estimate proportions for each of them.
+		// The absolute value of deploymentReplicasAdded should never exceed the absolute
+		// value of deploymentReplicasToAdd.
+		deploymentReplicasAdded := int32(0)
+		for i := range allRSs {
+			rs := allRSs[i]
+
+			proportion := getProportion(rs, deployment.Spec.Replicas, deploymentReplicasToAdd, deploymentReplicasAdded)
+
+			rs.Spec.Replicas += proportion
+			deploymentReplicasAdded += proportion
+		}
+
+		// Distribute any leftovers. The absolute value of deploymentReplicasAdded should never
+		// exceed the absolute value of deploymentReplicasToAdd based on the proportions
+		// estimation in getProportion above.
+		leftover := deploymentReplicasToAdd - deploymentReplicasAdded
+		for i := 0; i < int(math.Abs(float64(leftover))); i++ {
```

Comment:
Can you explain "safer"? I don't want to gate the count to a specific range
even if the most common case will be 1 or 2

On Tue, May 24, 2016 at 5:56 PM, Phillip Wittrock <notifications@github.com>
wrote:

> In pkg/controller/deployment/deployment_controller.go
> <https://github.com/kubernetes/kubernetes/pull/20273#discussion_r64419972>
> :
>
> > +		// value of deploymentReplicasToAdd.
> > +		deploymentReplicasAdded := int32(0)
> > +		for i := range allRSs {
> > +			rs := allRSs[i]
> > +
> > +			proportion := getProportion(rs, deployment.Spec.Replicas, deploymentReplicasToAdd, deploymentReplicasAdded)
> > +
> > +			rs.Spec.Replicas += proportion
> > +			deploymentReplicasAdded += proportion
> > +		}
> > +
> > +		// Distribute any leftovers. The absolute value of deploymentReplicasAdded should never
> > +		// exceed the absolute value of deploymentReplicasToAdd based on the proportions
> > +		// estimation in getProportion above.
> > +		leftover := deploymentReplicasToAdd - deploymentReplicasAdded
> > +		for i := 0; i < int(math.Abs(float64(leftover))); i++ {
>
> It is probably safer just to add the leftovers to the RS with the most
> pods and leave a comment. This number should almost always be very small (N
> / 2 - 1; N = RS count).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20273/files/43382ad2c805c6f8f7095f8df180a9a3b2341a84#r64419972>
>